Gems/Atom/Utils/Code/Include/Atom/Utils/ImGuiCpuProfiler.inl

```diff
@@ -251,5 +263,516 @@ namespace AZ
+                                const float pixelDeltaNormalized = deltaX / ImGui::GetWindowWidth();
+                                auto tickDelta = aznumeric_cast<AZStd::sys_time_t>(-1 * pixelDeltaNormalized * GetViewportTickWidth());
+                                m_viewportStartTick += tickDelta;
+                                m_viewportEndTick += tickDelta;
+
+                                ImGui::SetScrollY(ImGui::GetScrollY() + deltaY * -1);
+                            }
+                        }
+                        else if (io.MouseWheel != 0 && io.KeyCtrl) // Zooming
+                        {
+                            // We want zooming to be relative to the mouse's current position
+                            const float mouseVel = io.MouseWheel;
+                            const float mouseX = ImGui::GetMousePos().x;
+
+                            // Find the normalized position of the cursor relative to the window
+                            const float percentWindow = (mouseX - ImGui::GetWindowPos().x) / ImGui::GetWindowWidth();
+
+                            const auto overallTickDelta = aznumeric_cast<AZStd::sys_time_t>(0.05 * io.MouseWheel * GetViewportTickWidth());
+
+                            // Split the overall delta between the two bounds depending on mouse pos
+                            const auto newStartTick = m_viewportStartTick + aznumeric_cast<AZStd::sys_time_t>(percentWindow * overallTickDelta);
+                            const auto newEndTick = m_viewportEndTick - aznumeric_cast<AZStd::sys_time_t>((1-percentWindow) * overallTickDelta);
+
+                            // Avoid zooming too much, start tick should always be less than end tick
+                            if (newStartTick < newEndTick)
+                            {
+                                m_viewportStartTick = newStartTick;
+                                m_viewportEndTick = newEndTick;
+                            }
+                        }
+                    }
+                }
+                ImGui::EndChild();
+            }
+            ImGui::End();
+        }
+
+        inline void ImGuiCpuProfiler::CollectFrameData()
+        {
+            const RHI::CpuProfiler::TimeRegionMap& timeRegionMap = RHI::CpuProfiler::Get()->GetTimeRegionMap();
+
+            m_viewportStartTick = INT64_MAX;
+            m_viewportEndTick = INT64_MIN;
+
+            // Iterate through the entire TimeRegionMap and copy the data since it will get deleted on the next frame
+            for (const auto& [threadId, singleThreadRegionMap] : timeRegionMap)
+            {
+                // The profiler can sometime return threads without any profiling events when dropping threads, FIXME
+                if (singleThreadRegionMap.size() == 0)
+                {
+                    continue;
+                }
+
+                // Now focus on just the data for the current thread
+                AZStd::vector<TimeRegion> newData;
```

Comment:
You can use newData.reserve(size); to avoid reallocation if the size is known.

Reply:
I threw in a call to reserve(singleThreadRegionMap.size()) which will avoid resizing in the case that each region only has one invocation (which should be the normal case). If any region has more than one invocation though it's going to allocate which I think might be unavoidable here, since we can't know the total number of regions in the singleThreadRegionMap without traversing it.